src/riak_kv_ts_svc.erl

```diff
@@ -26,55 +26,87 @@
 
 -export([decode_query_common/2,
          process/2,
          process_stream/3]).
 
+-type ts_requests() :: #tsputreq{} | #tsdelreq{} | #tsgetreq{} |
+                       #tslistkeysreq{} | #tsqueryreq{}.
+-type ts_responses() :: #tsputresp{} | #tsdelresp{} | #tsgetresp{} |
+                        #tslistkeysresp{} | #tsqueryresp{} |
+                        #tscoverageresp{} |
+                        #rpberrorresp{}.
+-type ts_get_response() :: {tsgetresp, {list(binary()), list(atom()), list(list(term()))}}.
+-type ts_query_response() :: {tsqueryresp, {list(binary()), list(atom()), list(list(term()))}}.
+-type ts_query_responses() :: #tsqueryresp{} | ts_query_response().
+-type ts_query_types() :: ?DDL{} | riak_kv_qry:sql_query_type_record().
+-export_type([ts_requests/0, ts_responses/0,
+              ts_query_response/0, ts_query_responses/0,
+              ts_query_types/0]).
+
 decode_query_common(Q, Cover) ->
-    %% convert error returns to ok's, this means it will be passed into
-    %% process which will not process it and return the error.
-    case catch decode_query(Q, Cover) of
-        {ok, DecodedQuery} ->
-            PermAndTarget = decode_query_permissions(DecodedQuery),
-            {ok, DecodedQuery, PermAndTarget};
-        {error, Error} ->
+    case decode_query(Q, Cover) of
+        {QueryType, {ok, Query}} ->
+            {ok, Query, decode_query_permissions(QueryType, Query)};
+        {_QueryType, {error, Error}} ->
             {ok, make_decoder_error_response(Error)};
-        {'EXIT', {Error, _}} ->
+        {error, Error} ->
+            %% convert error returns to ok's, this means it will be passed into
+            %% process which will not process it and return the error.
             {ok, make_decoder_error_response(Error)}
     end.
 
 -spec decode_query(Query::#tsinterpolation{}, Cover::term()) ->
     {error, _} | {ok, ts_query_types()}.
-decode_query(#tsinterpolation{ base = BaseQuery }, Cover) ->
-    Lexed = riak_ql_lexer:get_tokens(binary_to_list(BaseQuery)),
-    case riak_ql_parser:ql_parse(Lexed) of
-        {select, SQL} ->
-            riak_kv_ts_util:build_sql_record(select, SQL, Cover);
-        {describe, SQL} ->
-            riak_kv_ts_util:build_sql_record(describe, SQL, Cover);
-        {insert, SQL} ->
-            riak_kv_ts_util:build_sql_record(insert, SQL, Cover);
+decode_query(#tsinterpolation{base = BaseQuery}, Cover) ->
+    case catch riak_ql_parser:ql_parse(
+                 riak_ql_lexer:get_tokens(  %% yecc can throw nasty 'EXIT' exceptions
+                   binary_to_list(BaseQuery))) of
         {ddl, DDL, WithProperties} ->
-            {ok, {DDL, WithProperties}};
-        Other ->
-            Other
+            {ddl, {ok, {DDL, WithProperties}}};
+        {QryType, SQL} when QryType /= error,
+                            QryType /= 'EXIT' ->
+            {QryType, riak_kv_ts_util:build_sql_record(QryType, SQL, Cover)};
+        {'EXIT', {Reason, _StackTrace}} ->
```

Comment:
You are correct. It would separate the 'good' branches from the ones catching error, and that argument, feeble as it is, has won.